crates/trie/sparse-parallel/src/trie.rs

```diff
@@ -1230,6 +1234,135 @@ impl ParallelSparseTrie {
+        for RevealedSparseNode { path, masks, .. } in nodes.iter() {
+            if let Some(tree_mask) = masks.tree_mask {
+                self.branch_node_tree_masks.insert(*path, tree_mask);
+            }
+            if let Some(hash_mask) = masks.hash_mask {
+                self.branch_node_hash_masks.insert(*path, hash_mask);
+            }
+        }
+
+        let mut nodes_slice = nodes.as_slice();
+
+        // Due to the sorting, if there are any nodes to be revealed in the upper trie then the
+        // first node will be an upper trie node. Handle all upper trie reveals using
+        // `ParallelSparseTrie::reveal_node`, which deals with revealing recursively into
+        // lower-subtries as necessary.
+        if SparseSubtrieType::from_path(&nodes[0].path).lower_index().is_none() {
+            let mut num_upper_nodes = 0;
+            let upper_nodes =
+                nodes.iter().take_while(|n| SparseSubtrieType::path_len_is_upper(n.path.len()));
+
+            for node in upper_nodes {
+                // TODO once reveal_node is not on the trait we can probably make it accept
+                // references, and skip these clones.
+                self.reveal_node(node.path, node.node.clone(), node.masks.clone())?;
+                num_upper_nodes += 1;
+            }
+
+            // truncate the upper trie nodes off the front of the nodes slice.
+            nodes_slice = &nodes_slice[num_upper_nodes..];
+        }
+
+        #[cfg(not(feature = "std"))]
+        // Reveal lower subtrie nodes serially if nostd
+        for node in nodes_slice {
+            for node in nodes {
+                self.reveal_node(node.path, node.node.clone(), node.masks.clone())?;
+            }
+        }
+
+        #[cfg(feature = "std")]
+        // Reveal lower subtrie nodes in parallel
+        {
+            use rayon::iter::{IndexedParallelIterator, IntoParallelIterator, ParallelIterator};
+
+            // Group the nodes by lower subtrie. This must be collected into a Vec in order for
+            // rayon's `zip` to be happy.
+            let node_groups: Vec<_> = nodes_slice
+                .chunk_by(|node_a, node_b| {
+                    SparseSubtrieType::from_path(&node_a.path) ==
+                        SparseSubtrieType::from_path(&node_b.path)
+                })
+                .collect();
+
+            // Take the lower subtries in the same order that the nodes were grouped into, so that
+            // the two can be zipped together. This also must be collected into a Vec for rayon's
+            // `zip` to be happy.
+            let lower_subtries: Vec<_> = node_groups
+                .iter()
+                .map(|nodes| {
+                    let node = nodes.first().expect("no empty groups");
+                    let Some(idx) = SparseSubtrieType::from_path(&node.path).lower_index() else {
+                        panic!("upper subtrie node {node:?} found amongst lower nodes");
```

Comment:
i think can be replaced with expect?
```suggestion
                    let idx = SparseSubtrieType::from_path(&node.path).lower_index().expect("upper subtrie node {node:?} found with lower nodes");
```

Reply:
I think `expect` doesn't allow a formatting parameter, and I didn't want to incur the formatting unless it was necessary, so I went with the `else`/`panic!`

Reply:
ah I see